Migrate reCAPTCHA support to Google reCAPTCHA Enterprise (v2 and v3)

derlin opened the pull request:
The RegistrationRecaptcha logic is old and targets the first version of Google reCAPTCHA, which is no longer supported.

This commit updates the logic to align with the new **Google reCAPTCHA Enterprise**, with support for both v2 and v3 captchas:

- v2: checkbox-based, visible captcha that is displayed on the page
- v3: score-based, invisible captcha that is hidden on the page, and based on the user's behavior and identity on other sites.

The old configuration (site key, secret key) is replaced with project ID, site key and API key.

Two new optional configurations are:

- a "v3" toggle: there is no easy way to deduce from a site key which  version is enabled. The user hence has to provide this information  to the widget.
- a "min threshold": this is the minimal score a captcha must return  for the registration to proceed. It should be between 0.0 and 1.0,  and is set to the sensible default of 0.7 - "medium high".

The `register.ftl` template is updated as well. Note that in case of invisible captcha, a call to the captcha API must be made to retrieve the captcha token before submitting the form.
This and the new theme attributes will require custom theme maintainer to udpate their code.

The documentation is also updated to reflect the changes.

*About the new reCAPTCHA*

The new reCAPTCHA version works in multiple stages:

1. on the frontend, a JSON script is loaded on the page, and:
   * with v2, an input (populated automatically) is added to the form, with the checkbox,
   * with v3, nothing is shown, so it is up to us to request a captcha   token before submitting the form. This can be done with a button and a callback.
2. on the backend, an assessment must be created and sent to google, along with the captcha token (this is where the API key is necessary),
3. the google assessment response must then be interpreted, to decide if the reCAPTCHA is valid. This entails checking the action name, the "valid" attribute and the score.

Closes: https://github.com/keycloak/keycloak/issues/16138

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -150,11 +145,21 @@
+
             </div>
         </form>
-        <script type="module" src="${url.resourcesPath}/js/passwordVisibility.js"></script>
```
Why has this script been removed?

derlin replied:
Definitely a mistake, good catch!

jonkoops commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
Thanks for taking the effort to update the docs.

jonkoops commented:
@andymunro since this PR updates the documentation, can I ask you to give that a pass?

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -112,17 +112,10 @@
                                class="${properties.kcLabelClass!}">${msg("passwordConfirm")}</label>
                     </div>
                     <div class="${properties.kcInputWrapperClass!}">
-                        <div class="${properties.kcInputGroup!}">
```
Looks like this `div` was removed, but should it have been?

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -112,17 +112,10 @@
                                    name="password-confirm"
                                    aria-invalid="<#if messagesPerField.existsError('password-confirm')>true</#if>"
                             />
-                            <button class="pf-c-button pf-m-control" type="button" aria-label="${msg('showPassword')}"
```
Same for this password reveal button, I don't think this should have been removed.

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -136,11 +129,13 @@
-                <div class="form-group">
-                    <div class="${properties.kcInputWrapperClass!}">
-                        <div class="g-recaptcha" data-size="compact" data-sitekey="${recaptchaSiteKey}"></div>
+                <#if recaptchaVisible!false>
```
Since this condition does not have another branch, I think we can roll this into the `recaptchaRequired` check, so there is only a single `if` statement checking both conditions.

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -150,9 +145,20 @@
-                    <input class="${properties.kcButtonClass!} ${properties.kcButtonPrimaryClass!} ${properties.kcButtonBlockClass!} ${properties.kcButtonLargeClass!}" type="submit" value="${msg("doRegister")}"/>
-                </div>
+                <#if recaptchaRequired?? && !(recaptchaVisible!false)>
+                    <script type="text/javascript">function onSubmitRecaptcha(token) { document.getElementById("kc-register-form").submit(); }</script>
```
For new scripts, we default to the [`module`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#applying_the_module_to_your_html) type to get some more [sane defaults](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#other_differences_between_modules_and_standard_scripts). Can we also format this so the JS is on multiple lines and indented?

derlin commented:
@jonkoops Thank you for the review! I rebased and updated register.ftl as requested. Let me know!

ssilvert reviewed (dismissed):
LGTM

pedroigor commented on services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java:
```diff
@@ -214,50 +292,17 @@ public FormAction create(KeycloakSession session) {
-        CONFIG_PROPERTIES.add(property);
-
-        property = new ProviderConfigProperty();
-        property.setName(USE_RECAPTCHA_NET);
```
Could you elaborate on why we are dropping support for using `recaptcha.net` as an alternative when there are constraints on using the `google.com` domain?

derlin replied:
Good catch! This was a misunderstanding on my part, I added this option back and updated the doc accordingly. Sorry for the mishap.

pedroigor reviewed (commented):
I would say we need something in release notes and migration guide as the changes here might break existing deployments.

@derlin I'm also curious about what you mean about the support level of reCAPTCHA. Do you mean this https://cloud.google.com/recaptcha-enterprise/docs/compare-versions? Or Google is dropping it completely and recommending using v2/v3?

ssilvert commented:
@derlin There are a couple of failing tests.  It looks like a couple of tests were not updated to reflect help text changes.  I think if you fix those then this will be ready to merge.

See https://github.com/keycloak/keycloak/actions/runs/6854683386/job/18638409570?pr=24223

jonkoops commented:
@derlin I believe we currently do no not have tests for this functionality. Would this be something you'd be interested in implementing this as well? Doesn't have to be part of this PR, we can handle it separately.

derlin commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -150,9 +150,23 @@
-                    <input class="${properties.kcButtonClass!} ${properties.kcButtonPrimaryClass!} ${properties.kcButtonBlockClass!} ${properties.kcButtonLargeClass!}" type="submit" value="${msg("doRegister")}"/>
-                </div>
+                <#if recaptchaRequired?? && !(recaptchaVisible!false)>
+                    <script type="text/javascript" type="module">
```
done!

jonkoops reviewed (dismissed):
LGTM

jonkoops commented:
@derlin could you squash the commits on your branch and sign off on the final commit (see [DTO check](https://github.com/keycloak/keycloak/pull/24223/checks?check_run_id=18699382265))?

derlin commented:
@jonkoops Conerning:
> @derlin I'm also curious about what you mean about the support level of reCAPTCHA. Do you mean this https://cloud.google.com/recaptcha-enterprise/docs/compare-versions? Or Google is dropping it completely and recommending using v2/v3?

To be honest, I always have a hard time parsing Google documentation, as it is so unclear. At first, I understood they only support enterprise for now on, but now looking again it seems I misunderstood. What I implemented is for Enterprise only, but there is still a simpler "v3" option available. The biggest changes are the script added to the HTML (`../enterprise.js` or `/api.js`), and the whole assessment thing with minimum scores are specific to enterprise... (the normal one is simpler).

(I feel so stupid right now).

So basically, I see multiple options:

1. Only support enterprise, which has the same *Free up to 1 million assessments per month* free tier (the only difference is you need to enable billing) - this PR,
2. Support both, and implement all into one class with a toggle (this may make the code quite complex)
3. Support both, in different classes. The non-enterprise v3 could still be the default (and needs to be updated), the Enterprise one in this PR can be used instead if the user choses to.

What do you think?

derlin commented:
@jonkoops any news on this?

jonkoops commented:
Sorry, I haven't had time to get back into a review. My backlog has been piling up a lot and we've been focusing on bug fixes. I'll see if I can make some time.

jonkoops reviewed (dismissed):
LGTM

derlin commented:
What are the next steps? Should I rewrite everything to support both recaptcha and recaptcha enterprise? Should I squash all my commits and add a proper DCO? Some guidance on how to get this PR moving would be nice.

pedroigor commented:
@derlin Yeah, please sign-off the commit and squash so we can run tests and run a 2nd round of review. 

IIRC, one of the key concerns was related to still supporting recaptcha. But I think you have fixed this?

derlin commented:
@pedroigor well, actually, this PR still **replaces** recaptcha with recaptcha enterprise. I am working on something to allow to support both, but this will require lots of changes (mostly in the documentation). What I propose is basically to have these settings:

<img width="565" alt="Screenshot 2024-01-05 at 19 46 45" src="https://github.com/keycloak/keycloak/assets/5463445/69c8185e-c643-4ac1-b2d8-fd6cbf07982f">

The project ID would be the switch between enterprise and regular. The min. score threshold is the only thing that is not shared (action is actually also used in normal (sorry for the error in the screenshot). The second field would be used for either the site secret or the google api key, depending on whether project id is set. Do you think this is ok from the user perspective? (there will be help texts for every option).

Should I rebase and squash/recreate all commits with this version or still add commits on top so reviewers can follow what is happening?

pedroigor commented:
@derlin Now I'm also confused. I remember testing your PR and having it working with recaptcha. At that time, IIRC my only concern was related to the `recaptcha.net` ...

Perhaps we should just have your initial proposal ready (both versions from the same action) and then evaluate. We will probably end up with distinct actions, I would say. But for now, let's make it work and then evaluate.

Yeah, multiple commits are fine for now.

Does it make sense for you?

derlin commented:
@pedroigor I pushed the proposition for supporting normal/enterprise and v2/v3. I havent updated the docs for now, waiting for a feedback. I don't know how you tested... What did you put as project ID and api keys? Normal recaptcha only provide two values (site key, secret key), while enterprise require three values (project id, site key and api key).

andymunro commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
```diff
@@ -6,45 +6,52 @@
 ----
 
-. Create an API key to get your reCAPTCHA site key and secret. Note the reCAPTCHA site key and secret for future use in this procedure.
+. Create a key for a "Website" platform, and choose the desired key type. Leave the defaults for v3 reCAPTCHA (invisible), or toggle *Use checkbox challenge* for a v2 reCAPTCHA (visible). Note the site key for future use in this procedure.
```
```suggestion
. Create a key for a "Website" platform, and choose the desired key type. Leave the defaults for v3 reCAPTCHA (invisible), or enable *Use checkbox challenge* for a v2 reCAPTCHA (visible). Note the site key for future use in this procedure.
```

andymunro reviewed (changes requested):
Looks good, @derlin. Just one change to suggest.

pedroigor commented:
@derlin Thanks, looking at it this week.

derlin commented:
> Looks good, @derlin. Just one change to suggest.

I still have to redo the documentation as I now incorporated normal and enterprise support. Waiting for @pedroigor to validate the new code before updating it (we need to clearly explain the new flags and the two "modes").

pedroigor commented:
Hey @derlin. I did not forget about it :)

It works fine and I can use both recaptcha and recaptcha enterprise without any issue.

However, the UX is far from ideal and quite confusing. Mainly in regards to how enterprise is enabled (based on the project id). The same goes for the score, it is easy to get confused and mess things up until you find the right fields and values you need to set.

That is my only concern.

I've updated your PR with a commit that introduces a separate provider for enterprise. Please, let me know if it works for you.

Sorry for losing your previous commit message and you not being added as the author. Please, feel free to change the commit to have you as the author as well as improve the message if you want to. I had to do that to be able to run tests.

pedroigor commented:
Hey @derlin, wdyt about the latest changes? Do they work for you?

derlin commented:
@pedroigor So sorry, these past few weeks have been hectic. I am working on it! I plan to rebase, edit some of your changes, squash and add a commit on top with the fixes to the documentation. Not sure I will be able to do all that before the weekend though. Sounds good?

pedroigor commented:
@derlin Not a problem at all. Sure, take your time. We have time until we can merge changes for Keycloak 24 (2 weeks).

derlin committed:
Update ReCAPTCHA and add support for ReCAPTCHA Enterprise

Closes #16138

Signed-off-by: Lucy Linder <lucy.derlin@gmail.com>

derlin commented:
@jonkoops  @andymunro sorry to ping you guys. I wanted to test my changes on the documentation, but impossible to build it. I always get the same `IllegalArgumentException` from the header-maven-plugin. Do you have any clue on how to fix it? I would like to avoid committing something I haven't reviewed properly.

```
mvn install -pl docs/documentation/header-maven-plugin -Pdocumentation
Failed to execute goal org.apache.maven.plugins:maven-plugin-plugin:3.4:helpmojo (header-goal) on project header-maven-plugin: Execution header-goal of goal org.apache.maven.plugins:maven-plugin-plugin:3.4:helpmojo failed.: IllegalArgumentException
```

derlin committed:
Update documentation for reCAPTCHA support

Signed-off-by: Lucy Linder <lucy.derlin@gmail.com>

derlin commented:
I rebased and cleaned up the documentation, adding a section about recaptcha enterprise for extra clarity. Let me know if this works properly on your end, but I believe we are nearly there 😊 .

pedroigor reviewed (dismissed):
If @andymunro is OK with the docs, LGTM.

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -84,9 +84,23 @@
+                        }
+                    </script>
+                    <div id="kc-form-buttons" class="${properties.kcFormButtonsClass!}">
+                        <button class="${properties.kcButtonClass!} ${properties.kcButtonPrimaryClass!} ${properties.kcButtonBlockClass!} ${properties.kcButtonLargeClass!} g-recaptcha" 
```
A button should always have an explicit `type` attribute, this is to prevent accidental form submission, since the default type is `submit`.

derlin replied:
Added `type="submit"` to the button.

pedroigor replied:
I would also avoid doing that now but as a follow-up.

jonkoops commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -84,9 +84,23 @@
+                    </div>
+                <#else>
+                    <div id="kc-form-buttons" class="${properties.kcFormButtonsClass!}">
+                        <input class="${properties.kcButtonClass!} ${properties.kcButtonPrimaryClass!} ${properties.kcButtonBlockClass!} ${properties.kcButtonLargeClass!}" type="submit" value="${msg("doRegister")}"/>
```
Let's make things consistent and use a `button` element here as well.

pedroigor replied:
@jonkoops This is now breaking a lot of tests. I would propose to have follow-ups to any improvement we want to do as a follow-up. Otherwise, we are losing a bit of focus here ...

@derlin Sorry for the back-and-forth, again. I just tested reverting this part to how it was before (using an input) and it should be enough to pass tests and finally have this one in once the tests are OK.

jonkoops reviewed (changes requested):
Just a couple of small nits

derlin commented on themes/src/main/resources/theme/base/login/register.ftl:
```diff
@@ -69,10 +69,10 @@
                 <div class="form-group">
                     <div class="${properties.kcInputWrapperClass!}">
-                        <div class="g-recaptcha" data-size="compact" data-sitekey="${recaptchaSiteKey}"></div>
+                        <div class="g-recaptcha" data-size="compact" data-sitekey="${recaptchaSiteKey}" data-action="${recaptchaAction}"></div>
```
yes it is :) see https://developers.google.com/recaptcha/docs/display#auto_render

jonkoops replied:
Good to know, thanks!

jonkoops reviewed (dismissed):
LGTM, thanks @derlin!

derlin commented:
@jonkoops What are the next steps? Can I still do something?

ssilvert reviewed (dismissed):
Looks like the wrong UI controls are being used for selecting `recaptcha.net` versus `recaptcha v3` versus `none`.

Only one choice should be allowed, right?  So that would mean we should use something like radio buttons or a dropdown.

andymunro commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
```diff
@@ -6,19 +6,20 @@
 ----
 
-. Create an API key to get your reCAPTCHA site key and secret. Note the reCAPTCHA site key and secret for future use in this procedure.
+. Create a new reCAPTCHA and choose between Challenge v2 (visible checkbox) or Score-based, v3 (invisible) to get your reCAPTCHA site key and secret. Note them down for future use in this procedure.
```
```suggestion
. Create a reCAPTCHA and choose between Challenge v2 (visible checkbox) or Score-based, v3 (invisible) to get your reCAPTCHA site key and secret. Note them down for future use in this procedure.
```

andymunro commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
```diff
@@ -29,22 +30,64 @@ NOTE: The localhost works by default. You do not have to specify a domain.
++
+NOTE: The localhost works by default. You do not have to specify a domain.
++
+. On your Google Cloud Project, go to *Credentials* and create a new API key.
```
```suggestion
. On your Google Cloud Project, go to *Credentials* and create an API key.
```

andymunro commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
```diff
@@ -29,22 +30,64 @@ NOTE: The localhost works by default. You do not have to specify a domain.
++
+NOTE: For better security, click on *edit api key* and add an API restriction to restrict the key to the *reCAPTCHA Enterprise API* only.
++
+. Navigate to the {project_name} admin console.
```
```suggestion
. Navigate to the {project_name} Admin Console.
```

andymunro commented on docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:
```diff
@@ -29,22 +30,64 @@ NOTE: The localhost works by default. You do not have to specify a domain.
+.. Enter the *Recaptcha Site Key* generated at the beginning of the procedure.
+.. Enter the *Recaptcha API Key* generated at the beginning of the procedure.
+.. Toggle **reCAPTCHA v3** according to your Site Key type: on for score-based reCAPTCHA (v3), off for challenge reCAPTCHA (v2).
+.. (Optional) Customize the *Min. Score Threshold* as you see fit: this is the minimal score a user should achieve on reCAPTCHA to be allowed to register, between 0.0 and 1.0 (see https://cloud.google.com/recaptcha-enterprise/docs/interpret-assessment-website#interpret_scores[interpret scores]).
```
```suggestion
.. (Optional) Customize the *Min. Score Threshold* as you see fit. Set it to the minimum score, between 0.0 and 1.0, that a user should achieve on reCAPTCHA to be allowed to register. See https://cloud.google.com/recaptcha-enterprise/docs/interpret-assessment-website#interpret_scores[interpret scores].
```

andymunro reviewed (changes requested):
A few minor changes, @derlin

derlin commented:
> Looks like the wrong UI controls are being used for selecting `recaptcha.net` versus `recaptcha v3` versus `none`.
> 
> Only one choice should be allowed, right? So that would mean we should use something like radio buttons or a dropdown.

I am not sure I understand. Recaptcha.net and v3 are two different things. recaptcha.net is whether to use google.com or recaptcha.net as the domain to load the script, while v2/v3 is the type of captcha to use. They are both independent settings. And what do you mean by "None"?

andymunro reviewed (approved):
Looks good @derlin. Thanks for the updates.

jonkoops commented:
Looks like we have some failing tests, @derlin could you investigate the cause?

jonkoops committed:
Fix up selector for submit button

Signed-off-by: Jon Koops <jonkoops@gmail.com>

pedroigor committed:
Reverting unrelated changes to templates

Signed-off-by: Pedro Igor <pigor.craveiro@gmail.com>

pedroigor reviewed (approved):
@jonkoops I'm reverting the changes that were breaking tests. If you agree, we can improve things as a follow-up.

Run another round of tests and looks good to me.

@derlin Please, let us know if it is working well for you now so we can finally merge this one ...

jonkoops commented:
Sure, as long as we log an issue to change that input back to a button I am fine with that 👍

pedroigor commented:
@derlin Thanks a lot for your time on this. Looking forward to more improvements in this area :)

alexagc commented:
Hi @pedroigor @jonkoops, this new implementation (and the older one) with the new user profile breaks profile validation, by  default the max length of unmanaged attributes its 2048 and sometimes the g-recaptcha-response exceed that number breaking the registration. One temporal solution its add that property to user profile and change the default length to one over 2400 (we dont see any higher number).
Maybe the solution its to ignore that property like the password and password confirm on profile validation (https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/forms/RegistrationUserCreation.java#L251)
If its needed i can make a pull request with the changes.

jonkoops commented:
@alexagc feel free to shoot us a PR, that is always more than welcome. Also make sure to log an issue describing the problem and link it to your PR. Thanks!
